### PR TITLE
[WB-4543] Add wandb.run.define_metric()

### PR DESCRIPTION
**Pull request by raubitsj**
https://wandb.atlassian.net/browse/WB-4543

Description
-----------

Add wandb.run.define_metric()

Design
-------
https://paper.dropbox.com/doc/CLI-API-add-run.define_metric--BEm98DroFga614g7tvrdo6IXAg-z9GIYShHF1sa9fTlBfxxz
(this design has evolved over time but it captures many of the design decisions which were considered)
This PR serves as the most up-to-date information about the change.

Proposed changes based on feedback
---------------------------------
- Support nested histories, '.' used as separator.  If the user has "."s in their metric they will have to escape them. See examples
- Support nested aggregations.  If specified, the leaf value (which is currently the last_value) becomes a dictionary.  See examples.
- Add "last" aggregation, if enabled it will change summary
from:
{"metric":{"is":{"nested": {"min": 3}}}
to:
{"metric":{"is":{"nested": {"min": {"last": 3}}}}
- "mean" aggregation should also log "num" or "count" to eventually support resume
- add more validation to define_metric name string.  check for invalid chars, etc
- add "none" to summary, so we can avoid even capturing it for some metrics (need to clear others when seen, not siimple merge
- add user intent to json metadata,  previously the code would not inform the frontend of the glob, instead it would match metrics as they were seen,  While this is useful to keep frontend more simple. it hides information that might be useful.  A key case is determining x-axis on grouped runs.  If the last run is used to find the preferred x-axis, and this run ended early without logging the metric. then the frontend wont know to apply the x-axis.  The "fix" though complicated, is so the frontend could know the intent was to glob. and would have to apply a similar glob logic, to decide how to handle this grouped view.   Any user intent record will have a flag indicating that they are intent.

Only required change is to make sure the metadata has a name that is properly escaped.  everything else is optional for now as this PR doesnt support nested histories (yet)

Examples
----------

1) Metric with dots
Metric name: "this.metric.has.dots"
```
wandb.define_metric("this\.metric\.has\.dots", ...)
wandb.define_metric("this.metric.has.dots", ...).  <== Possibly allow (LATER)
wandb.log({"this.metric.has.dots": 3})
```
aggregations will be:
```
"this.metric.has.dots.min"
"this.metric.has.dots.max"
```

metric in metadata will be: (MUST be escaped)
```
"m": [ {"1": "this\.metric\.has\.dots",
```

Note 1:
It would be good to have a warning at the end of the run if a metric did not match, this would indicate a problem to the user.
Note 2:
We might accept the non-escaped string and match both for nested dict and the flat string.   REEQUIREMENT: the actual metadata needs to be escaped properly -- this makes sure that there is no ambiguity in the data stored.   The overhead of this double check is low and it simplifies the user experience. only risk is that we will match a flat string even if the user only wanted the dict form to be matched, this seems like a very small risk. 

2) Nested Metric
```
wandb.define_metric("metric.is.nested, ...)
wandb.log(dict(metric=dict(is=dict(nested=3))))
```
when aggregations are not set summary will remain:
```
{"metric":{"is":{"nested": LAST}}}
```

when aggregations are set:
- {"metric":{"is":{"nested": {"min": 3}}}
-  {"metric":{"is":{"nested": {"last": 3}}}

metric in metadata will be:
```
"m": [ {"1": "metric.is.nested",
```

3) mixed
```
wandb.define_metric("metric.is\.nested, ...)
wandb.define_metric("metric.is.nested, ...).  <== Possibly allow (LATER)
wandb.log({"metric":{"is.nested": 3}})
```

metric in metadata will be:
```
"m": [ {"1": "metric.is\.nested",
```

Highlights
----------
- Aggregations like (min,max,mean,best) can be automatically added to the summary
- Integrations can specified their preferred step metric (and pass that info to the UI) eg "global_step" instead of _step
- wandb.log() code can better coexist with integration logged data by using the `step_sync` option (configured automatically by the integration (See: https://github.com/PyTorchLightning/pytorch-lightning/pull/5931)
- independent-variable metrics can be hidden from auto charts (`hide`)
- user now has a way to provide additional information about a metric (including optimization direction - to be used for auto sweeps).   But we could do more here like tag the metric with some metadata that this is a loss curve, etc which could help match the data with custom charts.  could also enforce types, or specify limitations (non-decreasing values)

Open Questions
----------------
- Is the UI / backend team ok with this protocol buffer json encoding (already being used for telemetry data).  It would be better for this to be structured data but synchronizing backend and cli changes are hard so this tries to create a format that is small, but fully forward/backward compatible.   Currently it is in the _wandb field of config, but it should go into the gorilla metadata field (when that is available).   If the backend team has a better proposal, it is very easy to change CLI to store it some other way.

Testing
-------

- unit tests 100% covered diff
- regression tests passing
- Prototype UI changes https://github.com/wandb/core/pull/6460
- PytorchLightning work
   - https://github.com/PyTorchLightning/pytorch-lightning/pull/5931
   - https://colab.research.google.com/drive/1W8Rg4ZEjISNPX8q7jyyQaltku7uwFccv?usp=sharing

TODO
------
- [x] ~~make sparklines better~~ Not in this PR (combine with settings work)
- [x] add tests
- [x] PROTOTYPE: get someone to modify the frontend to respect config["_wandb"]["x_axis"] Thanks @KyleGoyette 
- [x] get feedback on the api
- [x] ~~make sure that min,max,last are oneofs~~ They are not
- [x] figure out if we can add best, direction
- [x] implement proposal changes in design doc
- [x] add summary op mean
- [x] add hide
- [x] implement metric glob
- [x] remove MetricItem, (only one metric per record from now on)
- [x] implement auto_step logic
- [x] get tests working again
- [x] ~~add publish recorder fixture (steal from the other PR as this could be first)~~ not needed
- [x] add goal? to metric (add summary op best)
- [x] ~~add best_step summary op?~~ Not required in this PR, maybe or maybe later
- [x] add docstring
- [x] get coverage up
- [x] add telemetry
- [x] send new-style metric
- [x] test and check against NaN
- [x] Multiple calls to define_metric() merges metrics, unless specified overwrite? should we save this for later? (how do we make this clean in the api. reset=True?... i dunno.
- [x] Default to minimize goal if best is chosen
- [x] ~~handle nested metrics?~~ No nested metric support in this PR, should be no change to existing library
- [x] ~~Figure out if this could improve tensorboard_sync (a bit harder since tb_reader is aynchronous from user process)~~ nothing trivial, but possibly some extended usage of syncing metrics between two writers.  I think it will be better than what we currently do but it means either backend ssupport for out of order writers, or the CLI merging the namespaces -- which defining metrics might make possible with some extended functionality
- [x] Make sure frontend is happy with protobuf encoded data

Example
--------
```python
import math
import wandb
run = wandb.init(project="WB-4543")
run.define_metric("global_step", hide=True)
run.define_metric("*", step_metric="global_step", step_sync=True, summary="min,max,best,mean", goal="maximize")
for x in range(5, 100):
    run.log(dict(global_step=x))
    val = 100*math.exp(-(((x-30)/15)**2)) + 60*math.exp(-(((x-70)/20)**2))
    run.log(dict(val=val))
```

Output
-------
Screen Shot 2021-02-15 at 3 50 44 PM

Screen Shot 2021-02-15 at 3 55 11 PM

- https://wandb.ai/jeffr/WB-4543/runs/5j451wv0
- https://dashboard.featurepeek.com/peek/0jvhnmv4#/jeffr/WB-4543/runs/5j451wv0

Metric data sent to UI:
```json
{
 "_wandb": {
        "desc": null,
        "value": {
            "m": [
                {
                    "1": "global_step",
                    "6": [
                        2
                    ]
                },
                {
                    "1": "val",
                    "5": 1,
                    "6": [
                        1
                    ],
                    "7": [
                        1,
                        2,
                        3,
                        4
                    ],
                    "8": 2
                }
            ],
      }
 }
}
```
Sample function to decode x-axis for a given metric:
https://github.com/wandb/client/blob/feature/define-metric/tests/utils/parse_metrics.py

```python
def get_step_metric_dict(ml):
    """Get mapping from metric to preferred x-axis."""
    nl = [m["1"] for m in ml]
    md = {m["1"]: nl[m["5"] - 1] for m in ml if m.get("5")}
    return md
```

### CODE REVIEW

**Comment by vanpelt** [comment]
Whoa, this is cool!  I was thinking we may want to also support `run.define_metric("val_acc", xaxis="global_step")` or `run.define_metric("val_acc").set_xaxis("global_step")`

**Comment by borisdayma** [comment]
I love it! In particular the x-axis feature!
On my side it just needs to be defined globally for now.

But I like Chris idea, to which we could add `run.define_metric(['train_loss`, `val_loss`])`.

What happens for people who changed their x-axis in W&B interface?
Would it reset it or does it become the new "default" (when we clear settings)?

**Comment by raubitsj** [comment]
> I love it! In particular the x-axis feature!
> On my side it just needs to be defined globally for now.
> 
> But I like Chris idea, to which we could add `run.define_metric(['train_loss', 'val_loss'])`.
I didn't implement a list, but I did add a wildcard.
The reason why i didn't add a list, is it might be necessary to pass a list to support nested metrics.

> What happens for people who changed their x-axis in W&B interface?
> Would it reset it or does it become the new "default" (when we clear settings)?
The intent is that this would only change the default x-axis for a metric.

**Comment by raubitsj** [comment]
> Whoa, this is cool! I was thinking we may want to also support `run.define_metric("val_acc", xaxis="global_step")` or `run.define_metric("val_acc").set_xaxis("global_step")`

I messed with the usage, it is snow more consistent with our other methods as you recommended (using parameters instead of setters).

File: .codecov.yml [hunk 1]
```diff
@@ -1,5 +1,8 @@
 codecov:
-  require_ci_to_pass: yes
+  require_ci_to_pass: no
```
Comment (vanpelt) [inline .codecov.yml:2]: ???

File: tests/conftest.py [hunk 2]
```diff
@@ -473,3 +492,207 @@ def log_debug(caplog):
...
+        for m in metrics:
+            internal_sender._publish_metric(m)
+        for h in history:
+            internal_sender.publish_history(**h)
+        stop_backend()
+
+        ctx_util = parse_ctx(mock_server.ctx)
+        return ctx_util
+
+    yield fn
```
Comment (vanpelt) [inline tests/conftest.py:698]: HOT 🔥

File: wandb/sdk/internal/handler.py [hunk 3]
```diff
@@ -168,6 +175,86 @@ def _save_history(self, record: Record) -> None:
...
+            tot_key = k + ".tot"
+            num_key = k + ".num"
+            avg_key = k + ".mean"
+            tot = self._metric_track.get(tot_key, 0)
+            num = self._metric_track.get(num_key, 0)
+            tot += float_v
+            num += 1
+            self._metric_track[tot_key] = tot
+            self._metric_track[num_key] = num
+            self._consolidated_summary[avg_key] = tot / num
```
Comment (vanpelt) [inline wandb/sdk/internal/handler.py:223]: May not be needed but we might want to cast `tot` into a float.  There the unset case on line 217 which will be an integer and I imagine a user could pass in an int...

File: wandb/sdk_py27/wandb_run.py [hunk 4]
```diff
@@ -1769,6 +1775,95 @@ def _save_job_spec(self):
...
+    ):
+        """Define metric properties which will later be logged with `wandb.log()`.
+
+        Arguments:
+            name: Name of the metric.
+            step_metric: Independent variable associated with the metric.
+            step_sync: Automatically add `step_metric` to history if needed.
+            hide: Hide this metric from automatic plots.
+            summary: Specify aggregate metrics added to summary.
+                Supported aggregations: "min,max,mean,best"
```
Comment (vanpelt) [inline wandb/sdk_py27/wandb_run.py:1797]: It wasn't clear to me how `best` was any different than min or max.  We should probably say something here about **best** defaulting to min if goal isn't specified or just say the default goal is minimize..

File: wandb/sdk_py27/wandb_run.py [hunk 5]
```diff
@@ -1769,6 +1775,95 @@ def _save_job_spec(self):
...
+            summary_ops = []
+            valid = {"min", "max", "mean", "best"}
+            for i in summary_items:
+                if i not in valid:
+                    raise wandb.Error(
+                        "Unhandled define_metric() arg: summary op: {}".format(i)
+                    )
+                summary_ops.append(i)
+        if goal is not None:
+            valid_goal = {"minimize", "maximize"}
```
Comment (vanpelt) [inline wandb/sdk_py27/wandb_run.py:1847]: _nit:_ we're using **min** and **max** for summary but **minimize** and **maximize** here...  May want to also accept min and max

**Review by vanpelt — APPROVED** [review]
This is AWESOME @raubitsj.  The tests are SICK, can't wait to play with this.  We should likely define metrics / steps for tensorboard and keras automatically once this is out...  Great work!

File: .codecov.yml (hunk 1 above)
Comment (raubitsj) [reply .codecov.yml:2]: It means we get coverage data after py27 and py36 rather than waiting for windows and mac to finish.
And we get coverage info even if something fails, which is also useful to know.

**Review by shawnlewis — COMMENTED** [review]
I haven't reviewed the code, but after our discussion and the small adjustments we made I agree with the API and design here.
